Вывод привычки

SAWINBEAST opened the pull request:
Доставание привычки и привычек из БД после запроса клиента и вывод их ему в чат

SAWINBEAST committed:
Test

nothing

SAWINBEAST committed:
Реализация вывода привычек

Вывод созданных привычек пользователем из БД (их названия, типы, описания)

MAXon28 commented on LifeHabitTracker/LifeHabitTracker.BusinessLogicLayer/Impls/Habits/HabitService.cs:
```diff
@@ -12,11 +12,17 @@ internal class HabitService : IHabitService
+        /// <summary>
+        /// Список привычек пользователя
+        /// </summary>
+        private static IList<Habit> _clientHabits;
```
Зачем это выносить на уровень члена класса?

SAWINBEAST replied:
Потому, что это нечто постоянное. У клиента всегда есть список привычек.  И у меня `_clientHabits` используется в двух методах (хоть они и последовательны, но зачем внутри `GetHabitsAsync `или даже `PrepareClientHabit` создавать List `clientHabits`, а потом его возвращать. )

SAWINBEAST replied:
![image](https://github.com/SAWINBEAST/Life-Habit-Tracker/assets/115563963/25829aee-9aa8-4631-a7c0-adbdcc222660)

SAWINBEAST replied:
![image](https://github.com/SAWINBEAST/Life-Habit-Tracker/assets/115563963/950c0db6-9965-4b08-a1f6-8998bf2b906d)

SAWINBEAST replied:
Не исправил

SAWINBEAST replied:
После твоих последующих советов `_clientHabits ` уже не нужен.
Исправил.

MAXon28 commented on LifeHabitTracker/LifeHabitTracker.BusinessLogicLayer/Impls/Habits/HabitService.cs:
```diff
@@ -73,6 +79,35 @@ private static DbTimes PrepareTimesData(Habit habit)
+        /// <returns>Коллекция привычек пользовательского вида</returns>
+        private static IReadOnlyCollection<Habit> PrepareClientHabit(IReadOnlyCollection<DbHabits> selectedHabits)
+        {
+            List<Habit> habits = new List<Habit>();
```
Этот метод можно написать так:
` => selectedHabits
            .Select(x => new Habit
            {
                Name = x.Name,
                Description = x.Description,
                Type = x.IsGood == true ? "Хорошая" : "Плохая"
            })
            .ToArray();`

SAWINBEAST replied:
Исправил. 
Но тогда мне придется возвращать в уровень  Консоли не `List<Habit>`, а `IReadOnlyCollection<Habit>`. Но это в принципе нестрашно, так как из манипуляций у меня только `Count()` идёт.

SAWINBEAST replied:
Спасибо за Идею с Linq !

MAXon28 commented on LifeHabitTracker/LifeHabitTracker.BusinessLogicLayer/Impls/Habits/HabitService.cs:
```diff
@@ -73,6 +79,35 @@ private static DbTimes PrepareTimesData(Habit habit)
+                });
+            }
+
+            return habits;
```
Опять же - зачем тогда возвращать значение, если уже поле на уровне класса есть? Но вернее именно так.

SAWINBEAST replied:
Да, смотрю сейчас и не понимаю, почему я так сделал)
Видимо, хотел что-то еще реализовать, а потом отвлекся и забыл

MAXon28 commented on LifeHabitTracker/LifeHabitTracker.BusinessLogicLayer/Impls/Habits/HabitService.cs:
```diff
@@ -73,6 +79,35 @@ private static DbTimes PrepareTimesData(Habit habit)
 
         /// <inheritdoc/>
-        public IReadOnlyCollection<Habit> GetHabits() => new Habit[] { new Habit() };
+        public async Task<IList<Habit>> GetHabitsAsync(long chatId)
```
`public async Task<IReadOnlyCollection<Habit>> GetHabitsAsync(long chatId)`

SAWINBEAST replied:
Да, это вытекающее исправление из замечания выше. Исправил.

MAXon28 commented on LifeHabitTracker/LifeHabitTracker.Console/TelegramBot.cs:
```diff
@@ -138,6 +137,32 @@ await _habitService.AddHabitAsync(habit, chatInfo.ChatId)
+        /// <param name="cancellationToken">Токен отмены</param>
+        private async Task HandleViewHabitCommandAsync(Chat chat, CancellationToken cancellationToken)
+        {
+            var habitNames = await _habitService.GetHabitsAsync(chat.Id);
```
Просто habits

SAWINBEAST replied:
Исправил.
Да, так логичнее. Я же получаю не только имена привычек, но и остальную информацию о них

MAXon28 commented on LifeHabitTracker/LifeHabitTracker.Console/TelegramBot.cs:
```diff
@@ -138,6 +137,32 @@ await _habitService.AddHabitAsync(habit, chatInfo.ChatId)
+            var habitNames = await _habitService.GetHabitsAsync(chat.Id);
+            string message;
+
+            if (habitNames != null && habitNames.Count() > 0)
```
`habitNames.Count`

SAWINBEAST replied:
Исправил.
Странно. В документации у IROC нет метода Count(), только свойство. Но у меня , почему-то работает)) 
У List , который был до этого, тоже есть только свойство Count, ,без метода. И у ICollection, и у Object.

MAXon28 commented on LifeHabitTracker/LifeHabitTracker.Console/TelegramBot.cs:
```diff
@@ -138,6 +137,32 @@ await _habitService.AddHabitAsync(habit, chatInfo.ChatId)
+        private async Task HandleViewHabitCommandAsync(Chat chat, CancellationToken cancellationToken)
+        {
+            var habitNames = await _habitService.GetHabitsAsync(chat.Id);
+            string message;
```
Для множественной работы со строками используй StringBuilder.

SAWINBEAST replied:
Да, точно. У нас же string - immutable. И при изменении строки мы просто создаем новую с тем же именем. А StringBuilder позволяет нам динамически изменять одну и ту же строку.
Спасибо, что подсказал. 
Исправил.

SAWINBEAST committed:
Улучшение существующего кода

Работа над ошибками

SAWINBEAST reviewed (commented):
Исправлены давние замечания

SAWINBEAST committed:
Чтение конкретной привычки (в процессе)

Промежуточный итог работы над функционалом чтения определённой привычки.

SAWINBEAST committed:
Выдача конкретной привычки (в процессе)

- Прописал реализацию селектов репозиториев каждой из таблиц.
- Поработал Сиквел-функциями.
- Изменил некоторые названия методов и типы старых объектов.
- Почти сделал приведение данных к виду клиента из вида БД на уровне бизнес-логики.

SAWINBEAST committed:
Выдача конкретной привычки (готово)

Код реализован. Все 3 уровня прописаны, Обработка данных под вид клиента доделана. Не работает запрос к БД.

SAWINBEAST committed:
Процесс чтения конкретной привычки

Исправлена ошибка чтения данных из БД.
Исправлены сопутствующие ошибки по удалению контекста пользователя.
Добавлены информативные блоки в пользовательский интерфейс бота.
Не исправлена ошибка по чтению полученных данных из таблицы в HabitRepository

SAWINBEAST committed:
Вывод конкретной привычки

Исправлены предыдущие ошибки, найдены новые. Не создаётся объект DbHabits.

SAWINBEAST committed:
Процесс выдачи конкретной привычки

Проблема в HabitRepository найдена и исправлена. Появились новые ошибки в DaysRepo.

SAWINBEAST committed:
Выдача инфы по конкретной привычке

Функционал полностью реализован. Есть некоторые моменты, которые нужно зарефакторить с точки зрения архитектры и чистоты алгоритмизации. Но всё работает.
